### PR TITLE
aws: Create subnets for additional network CIDRs

### PR DESCRIPTION
**Pull request by hakman**
This will be needed for scale tests, to be able to use clusters with more than 500 nodes.

### CODE REVIEW

File: cmd/kops/create_cluster.go [hunk 1]
```diff
@@ -300,7 +299,7 @@ func NewCmdCreateCluster(f *util.Factory, out io.Writer) *cobra.Command {
 	cmd.RegisterFlagCompletionFunc("subnets", completeSubnetID(options))
 	cmd.Flags().StringSliceVar(&options.UtilitySubnetIDs, "utility-subnets", options.UtilitySubnetIDs, "Shared utility subnets to use")
 	cmd.RegisterFlagCompletionFunc("utility-subnets", completeSubnetID(options))
-	cmd.Flags().StringVar(&options.NetworkCIDR, "network-cidr", options.NetworkCIDR, "Network CIDR to use")
+	cmd.Flags().StringSliceVar(&options.NetworkCIDRs, "network-cidr", options.NetworkCIDRs, "Network CIDR(s) to use")
```
Comment (prateekgogia) [inline cmd/kops/create_cluster.go:302]: Do we need to make this change here as well - https://github.com/kubernetes/kops/blob/master/tests/e2e/kubetest2-kops/deployer/up.go#L112?

Comment (prateekgogia) [inline cmd/kops/create_cluster.go:302]: My suggestion would be to add a new flag called `additional-network-cidrs` to be backward compatible.

Comment (hakman) [reply cmd/kops/create_cluster.go:302]: It is already backwards compatible. Nothing changes from user point of view.

Comment (hakman) [reply cmd/kops/create_cluster.go:302]: I don't see any mention of `network-cidr` there.

Comment (prateekgogia) [reply cmd/kops/create_cluster.go:302]: Sorry, might be missing something here but won't going from `StringVar` to `StringSliceVar` require users to make a change to pass a slice of strings now?

Comment (prateekgogia) [reply cmd/kops/create_cluster.go:302]: IIUC, kubetest2 uses that interface which I linked earlier and we would need to provide this flag there as well.

Comment (hakman) [reply cmd/kops/create_cluster.go:302]: you can pass a slice as multiple `--network-cidr` flags or comma separated.
For example, with this change, one can just do:
```
--network-cidr=10.0.0.0/16 --network-cidr=10.1.0.0/16 --network-cidr=10.2.0.0/16 --network-cidr=10.3.0.0/16 --network-cidr=10.4.0.0/16 --network-cidr=10.5.0.0/16
```

Comment (justinsb) [reply cmd/kops/create_cluster.go:302]: I think we can pass them via the "passthrough" CreateArgs: https://github.com/kubernetes/kops/blob/9e0db3db74fbe5e435bbf7d1af805ab07f34bf6e/tests/e2e/kubetest2-kops/deployer/up.go#L126-L135

Comment (justinsb) [reply cmd/kops/create_cluster.go:302]: Right, it's a nice feature of the flags library!

File: upup/pkg/fi/cloudup/new_cluster.go [hunk 2]
```diff
@@ -745,10 +747,32 @@ func setupZones(opt *NewClusterOptions, cluster *api.Cluster, allZones sets.Stri
...
 		}
-		zoneToSubnetMap[zoneName] = subnet
+		zoneToSubnetsMap[zoneName] = append(zoneToSubnetsMap[zoneName], subnet)
 	}
 
-	return zoneToSubnetMap, nil
+	if len(opt.NetworkCIDRs) > 1 {
+		for i, cidr := range opt.NetworkCIDRs[1:] {
+			zoneName := opt.Zones[i%len(opt.Zones)]
+			subnetName := fmt.Sprintf("%s-%d", zoneName, i+1)
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/new_cluster.go:756]: I think this is pretty specific to AWS.  Also the way we round-robin around the zones is convenient, but might be surprising (along with the numbering).

Should we feature-flag this feature (and/or restrict to AWS only at first)?

File: tests/integration/create_cluster/complex/expected-v1alpha2.yaml [hunk 3]
```diff
@@ -40,17 +44,29 @@ spec:
...
+    type: Public
+    zone: us-test-1a
+  - cidr: 10.2.0.0/16
+    name: us-test-1a-2
+    type: Public
+    zone: us-test-1a
+  - cidr: 10.3.0.0/16
+    name: us-test-1a-3
+    type: Public
+    zone: us-test-1a
```
Comment (johngmyers) [inline tests/integration/create_cluster/complex/expected-v1alpha2.yaml:69]: I'm thinking this test should have a second zone. And we should have a private_complex test.

File: upup/pkg/fi/cloudup/new_cluster.go (hunk 2 above)
Comment (johngmyers) [reply upup/pkg/fi/cloudup/new_cluster.go:756]: One possibility would be to put the `-%d` suffix on the first subnet as well when `len(opt.NetworkCIDRs) > 1` but on AWS the primary CIDR of a VPC does have slightly different semantics.

**Comment by johngmyers** [comment]
/lgtm

Holding off on approval to give opportunity to add those additional test cases. I would be okay with this going in as-is.

File: upup/pkg/fi/cloudup/new_cluster.go (hunk 2 above)
Comment (hakman) [reply upup/pkg/fi/cloudup/new_cluster.go:756]: I agree about the slightly different semantics. This is why I preferred to have different names for additional network CIDRs.

File: tests/integration/create_cluster/complex/expected-v1alpha2.yaml (hunk 3 above)
Comment (hakman) [reply tests/integration/create_cluster/complex/expected-v1alpha2.yaml]: Agreed, done.

**Comment by justinsb** [comment]
Thanks @hakman 

We discussed in office hours, this is good to go I think, my concern was whether we should have a feature flag. @johngmyers agreed with you that we didn't, so...

/approve
/lgtm

**Pull request merged**